lib/mongoid/search_indexable.rb

```diff
@@ -0,0 +1,166 @@
+# frozen_string_literal: true
+
+module Mongoid
+  # Encapsulates behavior around managing search indexes. This feature
+  # is only supported when connected to an Atlas cluster.
+  module SearchIndexable
+    extend ActiveSupport::Concern
+
+    # Represents the status of the indexes returned by a search_indexes
+    # call.
+    #
+    # @api private
+    class Status
+      # @return [ Array<Hash> ] the raw index documents
+      attr_reader :indexes
+
+      # Create a new Status object.
+      #
+      # @param [ Array<Hash> ] indexes the raw index documents
+      def initialize(indexes)
+        @indexes = indexes
+      end
+
+      # Returns the subset of indexes that have status == 'READY'
+      #
+      # @return [ Array<Hash> ] index documents for "ready" indices
+      def ready
+        indexes.select { |i| i['status'] == 'READY' }
+      end
+
+      # Returns the subset of indexes that have status == 'PENDING'
+      #
+      # @return [ Array<Hash> ] index documents for "pending" indices
+      def pending
+        indexes.select { |i| i['status'] == 'PENDING' }
+      end
+
+      # Returns the subset of indexes that are marked 'queryable'
+      #
+      # @return [ Array<Hash> ] index documents for 'queryable' indices
+      def queryable
+        indexes.select { |i| i['queryable'] }
+      end
+
+      # Returns true if all the given indexes are 'ready' and 'queryable'.
+      #
+      # @return [ true | false ] ready status of all indexes
+      def ready?
+        indexes.all? { |i| i['status'] == 'READY' && i['queryable'] }
+      end
+    end
+
+    included do
+      cattr_accessor :search_index_specs
+      self.search_index_specs = []
+    end
+
+    # Implementations for the feature's class-level methods.
+    module ClassMethods
+      # Request the creation of all registered search indices. Note
+      # that the search indexes are created asynchronously, and may take
+      # several minutes to be fully available.
+      #
+      # @return [ Array<String> ] The names of the search indexes.
+      def create_search_indexes
+        return if search_index_specs.empty?
+
+        collection.search_indexes.create_many(search_index_specs)
+      end
+
+      # Waits for the named search indexes to be created.
+      #
+      # @param [ Integer ] interval the number of seconds to wait before
+      #   polling again (only used when a progress callback is given).
+      #
+      # @yield [ SearchIndexable::Status ] the status object
+      def wait_for_search_indexes(names, interval: 5)
```

Comment:
We should probably also document `names` param

Reply:
wups, good catch!